Doc/qa process testing and dev (#26)
thoth-tech/handbook: docs/processes/quality-assurance/testing-and-dev.md

# QA Process - Testing and Development

> [Testing and Development](#testing-and-development

Reviewer: **issue:** Missing closing bracket?

- [Feature Development: where to start?](#feature-development-–-where-do-i-start)

- [Test-Driven Development](#test-driven-development)
  - [Writing Appropriate Tests](#determining-appropriate-tests)
  - [Writing a Failing Test Case](#writing-a-failing-test-case)
  - [Writing Code to Pass a Test](#writing-code-to-pass-your-test)
  - [Refactoring Code](#refactoring-code)
- [Epics and User Stories](#epics-and-user-stories)
  - [Epics](#epics)
  - [User Stories](#user-stories)
- [Testing Templates](#testing-templates)

Current Areas under construction:

- [Refactoring Intro](#tbc-refactoring-intro)
- [Waiting epic Template link](#tbc-template-link)
- [Testing tools expansion](#tbc-testing-tools)

# Testing and Development

Testing should be built into the software development process.

> Thoth-Tech advocates for a [Test Driven Development](#test-driven-development) approach.

Reviewer: **suggestion:** We could introduce the TDD acronym in the first use here. What do you think?

```suggestion
Thoth-Tech advocates for a [Test Driven Development](#test-driven-development) (TDD) approach.
```

But you might be wondering, before jumping into coding and planning tests for a new feature, _where do you begin_?

## Feature development – where do I start?

- Create [epics](#epics) for each feature for planned development
- Identify [user stories](#user-stories) of the feature

> - Consider how best to mitigate risk or look for alternate solutions

Reviewer: **suggestion:** I think it is really important to articulate the problem, or why the changes need to be made. I suggest we add an extra clause to stress the importance of that step.

```
- Clearly articulate the problem and why the changes need to be made
```

  - What's in/out of scope?

For each product/feature, using the required specifications, flesh out the expected behaviour for different use case scenarios. A collaborative approach helps to make sure you are defining the right problem and solution; be sure to seek feedback from the team during this process.

Armed with your acceptance criteria and use cases, you are ready to create the test scenarios that drive a [test driven development](#test-driven-development) approach.

You will also need to consider _how_ you are going to test your scenarios as part of this process – considering matters such as what tools will you use, what platforms tests will be performed on, any release processes, and what deliverables will you need to produce. This should be considered as part of your Test Strategy – a deliverable which should be produced as part of the QA process. Other documents that may be produced as part of the QA process include a Test Plan and Bug Reports.

Templates have been provided for sample [Test Strategy](templates/test-strategy-template.md), [Test Plan](templates/test-plan-template.md) and [Bug Report](templates/bug-report-template.md) documents.

> While manual testing may be conducted initially, moving towards automated tests allow for better efficiency, ease of code checks and is the preferred method to allow automatic test incorporation into the deployment pipeline. Testing and processes, like many other in the Agile model, are not static, and can (and should) be improved on successive iterations. This may include iterating existing processes for testing release as well as automating more functionality.

Reviewer: ```suggestion
While manual testing may be conducted initially, moving towards automated tests allows for better efficiency, ease of code checks and is the preferred method to allow automatic test incorporation into the deployment pipeline. Testing and processes, like many other in the Agile model, are not static, and can (and should) be improved on successive iterations. This may include iterating existing processes for testing releases as well as automating more functionality.
```

> ## **How do I prove my code does what it's supposed to do?**

Reviewer: **suggestion:** Could we drop the emphasis styles on headers 👇🏼 ? I don't think they add any visual difference.

Author: Ah... So it seems this _is_ making a visual difference on (at least some) headers - on my VS preview, anyway. The headings are only a little larger, but not bold without it - second image is removing the emphasis. Smaller headings especially seem to get lost a bit without it to my eye? @tancnle 
<img width="468" alt="image" src="https://user-images.githubusercontent.com/68226216/161381299-72ec1a78-3780-4d56-b4d8-a3fa0a7a589a.png">
<img width="468" alt="image" src="https://user-images.githubusercontent.com/68226216/161381313-0177f0cb-1e6f-4c41-946b-6000ca5a571d.png">

When you are satisfied that your code is complete, it is wise to challenge your test again with known failing code to reaffirm your testing is still functioning as expected.

**What next?**

### TBC refactoring intro \*\*\*

_Below block to be cleaned up/expanded_

```text
Code quality:

Is this reusable?

Is this understandable?

Is this readable?

Can a third-party looking at your code easily understand what is going on?

expand and intro into "refactoring code"
```

### **Refactoring Code**

Refactoring involves changing the _internal_ structure to make it more easily understood, effective and/or cheaper to modify without changing the _external behaviour_ of the code.

While writing a failing test case and then coding it to pass focuses on **requirements** (and adding **new functionality** ), refactoring focuses on **improving design**.

**My code is complete and passing tests; is it time to refactor?**

You don't need to assess your code for refactoring on every test, but if your find your code is:

- repeating itself (or repeating functionality),
- confusing to follow, or
- you are getting stuck,

these are good signs that it might be time to refactor.

> **How do I refactor?**

Reviewer: **praise:** Great addition to the development guide ❤️

Extract function refactoring is one common and intuitive approach to refactoring code: this involves comparing the code intention with the implementation. If the code is doing more than the intention, particularly if it is doing things that may be repeated elsewhere, it probably contains functions that should be split off.

[Refactoring.com](http://refactoring.com/) provides useful reading that covers a number of refactoring techniques.

**Further information about Test Driven Development**

Deakin students have free access to LinkedIn Learning courses.

> A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things?u=2104084)

Reviewer: **suggestion:** I think we can drop the `u` parameter as it is uniquely different for every user.

```suggestion
A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things).
```

Author: Thanks for picking that up, @tancnle !

> Developers working with C++ may also find [Test Driven Development in C++](https://www.linkedin.com/learning/test-driven-development-in-c-plus-plus/welcome?autoplay=true&u=2104084) useful

Reviewer: ```suggestion
Developers working with C++ may also find [Test Driven Development in C++](https://www.linkedin.com/learning/test-driven-development-in-c-plus-plus) useful.
```

- Drives creative solutions: the focus on the end-user solution required, rather than the how, encourages creative problem-solving (where to-do lists do not).

In Agile methodology, during a sprint or iteration, the team decides which user stories to tackle that story and discuss requirements; once agreed, requirements are added to the stories; Story tasks and subtasks can be tracked and assigned using the team's relevant Trello board; user stories for future sprints can remain in the backlog.

When defining user stories, consider:

- Definition of "done": how will we know when a task/story is complete?
- Tasks or subtasks: what steps need to be completed and by whom?
- User personas: who is the user being considered; are there multiple users (consider making user stories for each of them)
- Ordered steps: In a larger process, write a story for _each step_
- Time: consider how long a story may take; if it's more than one sprint, or seems very complex, the story may need to be broken down into smaller stories (or, you may be looking at an epic)

Further background on user stories can be found at https://www.atlassian.com/agile/project-management/user-stories

## Testing Templates

> [Sample Test Strategy](templates/test-strategy-template.md))
> 
> [Test Plan Template](templates/test-plan-template.md)
> 
> [Bug Report Template](templates/bug-report-template.md)

Reviewer: **nitpick:** We can convert this to a bullet lists

```suggestion
- [Sample Test Strategy](templates/test-strategy-template.md))
- [Test Plan Template](templates/test-plan-template.md)
- [Bug Report Template](templates/bug-report-template.md)

```

Author: Done - with a new intro sentence, as indented bullets looked a little off as the only thing under a heading.
